lib/internal/modules/run_main.js

```diff
@@ -92,10 +93,24 @@ function shouldUseESMLoader(mainPath) {
  */
 function runMainESM(mainPath) {
   const { loadESM } = require('internal/process/esm_loader');
-  const { pathToFileURL } = require('internal/url');
+  const { pathToFileURL, URL } = require('internal/url');
+  const _importMapPath = getOptionValue('--experimental-import-map');
```

Comment:
It's hardly a path, right?
```suggestion
  const _importMapReferrer = getOptionValue('--experimental-import-map');
```

Reply:
I meant "specifier", not "referrer", sorry. "specifier" is what the ECMA-262 uses to describe e.g. the first "argument" of dynamic imports.

Reply:
Why wouldn’t it be a path? It’s a file.

Reply:
Does that mean users would be forced to pass a file URL rather than a simple relative path?

Reply:
They can use relative URL strings still per https://github.com/nodejs/node/pull/50590/files#diff-9cdd961b23c4d25cc188f6b0beb3bf0228e3696a5f830cef198b1673b547bed3R106 but it would also do a full resolve at least in this PR so it would mean bare imports would work but not use import maps during the bootstrap.

Reply:
Yeah exactly as @bmeck describes. Honestly this was a late change recommended by @aduh95 @GeoffreyBooth and @JakobJingleheimer in the preliminary PR on my fork. I am on the fence on it personally, but I can see it might have value. I will add it to my list at the top of open discussion topics as well.

Reply:
Strongly agree with this! All of the tests are just using path's fyi: https://github.com/nodejs/node/pull/50590/files#diff-7421eeb82993796e3d7a6961ae584673dc7bbbc0fc46ab094ed9c8b8478ac7b6R155-R161

I honestly struggled while I was implementing this to get into the `URL` mode and I don't think they are ergonomic for most folks so I would hate to see too much of those semantics bleed out. Ideally we well support folks who just think about them all as paths first despite the internals all being `URL` like things per the spec.

Reply:
True, although it implies that anything other than bare specifiers should be "url like"

> "URL-like" specifiers, i.e., those that are either parseable as absolute URLs or start with "/", "./", or "../", can be remapped as well

From an implementation standpoint I actually did initially use path strings in my first iteration. @aduh95 recommended that I think about how to better use `URL`'s and it did simplify a bunch of the details of this PR. That said, I do specifically test bare specifiers for `imports` and `scopes`. But we have the open question of remapping *to* a bare specifier, which I removed from my initial implementation. I really think this is a feature we should support, but there was some question so I just opened the PR without it.

Reply:
For the content of the import maps: I'm very much doubtful that using paths would be what folks expect, e.g. let's take the following map:

```json
{
  "imports": {
    "foo": "./foo.js?key=value",
    "bar": "./my%20file.js"
  },
  "scopes": {}
}
```

If I `import 'foo'`, I would expect the above to fetch `foo.js`, and not have the querystring be interpreted as part of the filename.
Similarly, if I `import 'bar'`, I would expect that `my file.js` is imported, as would happen in a browser.

---

Regarding how should `--import-map=foo/bar` be interpreted as `./foo/bar` or `./node_modules/foo/bar.js`: I think it will depend if using static import maps from packages is a common use case. Using the same resolution algorithm is at least something easy to explain, and simplifies the implementation, that seems a good place to start.

Reply:
It has to be specifiers (URLs). This feature should be able to work with `--experimental-network-imports`, and the only way that’s possible is if they’re URLs.

It would also be terrible for interoperability if Node used paths while browsers use URLs. You should be able to use the same import map in Node as in a browser, at least when everything in the map is relative.

Reply:
I’m not talking about the import map contents, to be clear, just the relative files referenced on the command line. It’s fine if the map contains only URLs.

Reply:
> e.g. let's take the following map:

Honestly I think the query param example is exactly why I think URLs will be confusing to end users. The `bar` example even will likely be super confusing to an end user who opens up their import map imo. I understand why this is necessary, but I think it is an unfortunate necessity.

> I think it will depend if using static import maps from packages is a common use case

This is the only use case I strongly care to support. All the other use cases I see are for "toy" projects, experiments, or local dev UX stuff. Anything other than a static generated import map for production applications is IMO a bad idea to be discouraged. I recognize that I am probably in the minority in this opinion, so expect to make some concessions which do not agree with my personal opinion.

> Regarding how should --import-map=foo/bar be interpreted as ./foo/bar or ./node_modules/foo/bar.js

So with the above in mind, the current implementation will interpret it as the later (`node_modules`) one, and I am not a fan of that behavior but did it because using the standard resolution approach enables support for other features.

> You should be able to use the same import map in Node as in a browser

FWIW, I don't consider this a goal. That said, I think most common use cases should work, and being compatible where it makes sense *should* be a goal. I know it is a subtle difference, but if there is a trade-off between compatibility and UX I would fight for UX almost every time (and that aligns with our technical priorities).

That said, I do agree with "if Node used paths while browsers use URLs" for inside the import map.

Reply:
I changed the variable name to `specifier`. I think we have addressed the things from this thread. They *are* processed as `URL`s and so http specifiers and data uri's work, as will any other supported protocols. That said, we also are not going to require people to pass `--experimental-import-map=file://....` as that would just be bad UX for the most common use case. We use the `cwd` and resolve those specifiers as relative/absolute file urls.

If anyone thinks this is not a satisfactory answer we can re-open this thread.